### PR TITLE
story #351 Add/Get callback details for participants

### PR DESCRIPTION
**Pull request by shashi165**
**Coverage summary** 
Statements   : 53.83% ( 1236/2296 )
Branches     : 37.66% ( 177/470 )
Functions    : 39.28% ( 174/443 )
Lines        : 54.56% ( 1233/2260 )

1..402
tests 402
 pass  402

 ok

### CODE REVIEW

**Comment by ggrg** [comment]
GET http://{{BASE_URL_3001}}/participants/dfsp1/endpoints
resulted in 
{
    "id": "BadRequestError",
    "message": "Bad Request"
}
POST  http://{{BASE_URL_3001}}/participants/dfsp1/endpoints
with
{ 
	"endpoint": {
		"type": "FSIOP_CALLBACK_URL",
		"value": "abc"
	}
	
}
also resulted in 
{
    "id": "BadRequestError",
    "message": "Bad Request"
}

Reason:
src/models/participant/participant.js:45
err.message = "Cannot read property 'findOne' of undefined"

**Comment by ggrg** [comment]
Please, post coverage report.
Please, create pull requests from your fork.
Also, not sure if the object for the POST is correct. @rmothilal ?

**Comment by ggrg** [comment]
# Coverage summary
Statements   : 53.44% ( 1218/2279 )
Branches     : 34.82% ( 156/448 )
Functions    : 39.41% ( 175/444 )
Lines        : 54.17% ( 1215/2243 )

1..397
tests 397
pass  397

**ok**

**Comment by ggrg** [comment]
1. GET http://{{BASE_URL_3001}}/participants/dfsp1/endpoints throws error if endpoint does not exist
2. GET by type and GET all return different objects:
a. http://{{BASE_URL_3001}}/participants/dfsp1/endpoints?type=FSIOP_CALLBACK_URL
{
    "endpoints": {
        "type": "FSIOP_CALLBACK_URL",
        "value": "abcd"
    }
}
b. http://{{BASE_URL_3001}}/participants/dfsp1/endpoints
{
    "endpoints": [
        {
            "type": "FSIOP_CALLBACK_URL",
            "value": "abc"
        }
    ]
}
3. Subsequent POST http://{{BASE_URL_3001}}/participants/dfsp1/endpoints with
{ 
	"endpoint": {
		"type": "FSIOP_CALLBACK_URL",
		"value": "abcd"
	}
}
returns 201 Created
and value is updated.

File: src/domain/security/permissions.js [hunk 1]
```diff
@@ -13,6 +13,8 @@ const permissions = {
   PARTICIPANTS_LIST: new Permission('PARTICIPANTS_LIST', 'List all participants'),
   PARTICIPANTS_VIEW: new Permission('PARTICIPANTS_VIEW', 'View participant details'),
   PARTICIPANTS_UPDATE: new Permission('PARTICIPANTS_UPDATE', 'Update participant'),
+  PARTICIPANTS_ENDPOINTS_ADD: new Permission('PARTICIPANTS_ENDPOINTS_ADD', 'Add/Update participant endpoints'),
```
Comment (rmothilal) [inline src/domain/security/permissions.js]: what is this new permissions that you added? why did you add this?

File: src/models/participant/facade.js [hunk 2]
```diff
@@ -47,6 +51,130 @@ const getByNameAndCurrency = async (name, currencyId) => {
...
+ */
+
+const getAllEndpoints = async (participantId) => {
+  try {
+    return Db.participantEndpoint.query(builder => {
+      return builder.innerJoin('endpointType AS et', 'participantEndpoint.endpointTypeId', 'et.endpointTypeId')
+        .where({
+          'participantEndpoint.participantId': participantId,
+          'participantEndpoint.isActive': 1
+        }).select('participantEndpoint.participantEndpointId',
```
Comment (rmothilal) [inline src/models/participant/facade.js]: these are all the columns could you not just use .select('participantEndpointId.*', 'et.name')

File: src/models/participant/facade.js [hunk 3]
```diff
@@ -47,6 +51,130 @@ const getByNameAndCurrency = async (name, currencyId) => {
...
+
+const getEndpoint = async (participantId, endpointType) => {
+  try {
+    return Db.participantEndpoint.query(builder => {
+      return builder.innerJoin('endpointType AS et', 'participantEndpoint.endpointTypeId', 'et.endpointTypeId')
+        .where({
+          'participantEndpoint.participantId': participantId,
+          'participantEndpoint.isActive': 1,
+          'et.name': endpointType
+        }).select('participantEndpoint.participantEndpointId',
```
Comment (rmothilal) [inline src/models/participant/facade.js]: these are all the columns could you not just use .select('participantEndpointId.*', 'et.name')

File: testPI2/unit/shared/setup.test.js [hunk 4]
```diff
@@ -37,7 +37,7 @@ Test('setup', setupTest => {
     sandbox.stub(RequestLogger, 'logRequest')
     sandbox.stub(RequestLogger, 'logResponse')
 
-    Sidecar.connect = sandbox.stub()
+    sandbox.stub(Sidecar, 'connect')
```
Comment (rmothilal) [inline testPI2/unit/shared/setup.test.js]: why did you change this from Sidecar.connect = sandbox.stub()?

File: testPI2/unit/shared/setup.test.js [hunk 5]
```diff
@@ -95,6 +95,7 @@ Test('setup', setupTest => {
         test.ok(Sidecar.connect.calledWith(service))
         test.notOk(Migrator.migrate.called)
         test.equal(s, server)
+        Sidecar.connect.restore()
```
Comment (rmothilal) [inline testPI2/unit/shared/setup.test.js]: why did you change this from Sidecar.connect = sandbox.stub()?

Comment (shashi165) [reply testPI2/unit/shared/setup.test.js]: This was causing the tests to fail in admin/participants/handler.test.js .. since the Sidecar.connect was not being restored

File: testPI2/unit/shared/setup.test.js (hunk 4 above)
Comment (shashi165) [reply testPI2/unit/shared/setup.test.js]: This was causing the tests to fail in admin/participants/handler.test.js .. since the Sidecar.connect was not being restored

**Comment by shashi165** [comment]
merged the changes for #351 successfully
